### PR TITLE
WIP feat(client): complete on fennec

### LINKED ISSUE #4372: Add the new Complete Sync setup page
Ref: https://github.com/mozilla/fxa-content-server/pull/4370

image

### PR DESCRIPTION
**Pull request by shane-tomlinson**
Builds on #4395 

Fixes #4372

### CODE REVIEW

**Comment by rfk** [comment]
:heart:

One question I had when I saw this demo was around PII handling.  IIUC this will technically share your email address with our SMS provider, if they wanted to go looking for it in the links we're sending out.  If so then we should parallelize and seek data-approval at the same time as we're working on the code here.

(I'm having flashbacks to Gravatar where we had to add a permission box to even send them the *hash* of your email address)

**Comment by shane-tomlinson** [comment]
The title was misnamed, this is only "open verification link in fennec and allow the user to sign in". SMS is a much larger project that will not be part of this PR. 

We are developing this portion independently so that it can merge w/o being blocked on SMS work.

File: app/scripts/templates/connect_this_firefox.mustache [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  <section>
+    <div class="graphic graphic-connect-another-device">{{#t}}Success{{/t}}</div>
+
+    <p>
+      {{#t}}Complete your set-up by connecting this Firefox.{{/t}}
+    </p>
+
+    <form novalidate>
+      <div class="button-row">
+        <a class="button" href="{{{ escapedSignInUrl }}}">{{#unsafeTranslate}}Continue &rarr;{{/unsafeTranslate}}</a>
```
Comment (vladikoff) [inline app/scripts/templates/connect_this_firefox.mustache]: please add a class or an id to this button

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: bf5a8541b28560cafc9488022beb75a8889c081d
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/mozilla/fxa-content-server/blob/master/CONTRIBUTING.md#git-commit-guidelines

---
This message was auto-generated by https://gitcop.com

**Comment by vladikoff** [comment]
https://github.com/mozilla/fxa-content-server/pull/4395 landed so we can now use UAP for this

**Comment by vladikoff** [comment]
@shane-tomlinson is this waiting for the feature doc phase 1?

**Comment by shane-tomlinson** [comment]
> @shane-tomlinson is this waiting for the feature doc phase 1?

It's waiting on that, and on time from me. I'm pulling parts out of this to merge independently, will have a PR later.

File: tests/functional/lib/helpers.js [hunk 2]
```diff
@@ -1265,6 +1290,17 @@ define([
     };
   }
```
Comment (shane-tomlinson) [inline tests/functional/lib/helpers.js]: Add a JSDoc to `testUrlInclude`

File: app/scripts/views/connect_another_device.js [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+     * @private
+     */
+    _getEscapedSignInUrl (context, email) {
+      const origin = this.window.location.origin;
+
+      const params = {
+        context,
+        email,
+        service: Constants.SYNC_SERVICE,
+        utm_source: View.UTM_SOURCE //eslint-disable-line camelcase
```
Comment (shane-tomlinson) [inline app/scripts/views/connect_another_device.js]: Maybe a better option is to pass all the current utm_* query parameters verbatim, and use `entrypoint` instead of `utm_source`

**Comment by shane-tomlinson** [comment]
Still TODO:

* [ ] - Hook up "Why is this required?" - https://github.com/mozilla/fxa-content-server/issues/4485
* [ ] - Ensure the image shrinks as needed on mobile browsers
* [ ] - Hook up metrics for users that are able to sign in on 2nd device and those that click "sign in"

screen shot 2016-12-01 at 15 33 36

@vladikoff - this is beginning to be in reasonable shape, mind starting the review process? The PR is large, so I fully expect it to take a few rounds. I tried breaking off several sub-tasks into separate PRs to simplify review, this is what's left. I might be able to extract another PR or two if you think that'll make review easier.

File: app/scripts/lib/constants.js [hunk 4]
```diff
@@ -112,8 +112,13 @@ define(function (require, exports, module) {
     // Login delay for iOS broker
     IOS_V1_LOGIN_MESSAGE_DELAY_MS: 10000,
 
-    UNBLOCK_CODE_LENGTH: 8
+    UNBLOCK_CODE_LENGTH: 8,
+
+    MARKETING_ID_SPRING_2015: 'spring-2015-android-ios-sync',
+    MARKETING_ID_AUTUMN_2016: 'autumn-2016-connect-another-device',
+
+    DOWNLOAD_LINK_ANDROID: 'https://app.adjust.com/2uo1qc?campaign=fxa-conf-page&adgroup=android',
```
Comment (vladikoff) [inline app/scripts/lib/constants.js]: The problem with adding this to `constants.js` is the fact that it has `campaign=fxa-conf-page`, we need to confirm that we want all download links to have that campaign. *Shrugs*

File: app/scripts/templates/connect_another_device.mustache [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    {{/canSignIn}}
+    {{^canSignIn}}
+      {{#isFirefoxIos}}
+        <p id="signin-fxios">
+          {{#t}}Complete set-up by signing in to this Firefox.{{/t}}
+          {{#t}}Open settings and click Sign In to Firefox.{{/t}}
+        </p>
+      {{/isFirefoxIos}}
+
+      {{#isOtherIos}}
```
Comment (vladikoff) [inline app/scripts/templates/connect_another_device.mustache]: nit: `isOtherIos` sounds strange, maybe `isNotFirefoxIos` ?

Comment (vladikoff) [inline app/scripts/templates/connect_another_device.mustache]: but I guess that would mean any other UA. hmm

File: app/scripts/views/connect_another_device.js [hunk 6]
```diff
@@ -0,0 +1,200 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/**
+ * If the user verifies their email in an instance of Firefox
+ * that other than the one they used to sign up, suggest
+ * that they sign in.
```
Comment (vladikoff) [inline app/scripts/views/connect_another_device.js]: `suggest that they sign in.` or also download app?

**Comment by vladikoff** [comment]
`app/images/account-verified@2x.png` can be optimized and size reduced by 55% (tested via https://tinypng.com/)

File: app/scripts/views/connect_another_device.js [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+      'click #signin': '_onSignInClick'
+    },
+
+    beforeRender () {
+      return this.getAccount().isSignedIn()
+        .then((isSignedIn) => {
+          this.model.set('isSignedIn', isSignedIn);
+
+          this.notifier.trigger(`connectAnotherDevice.signedin.${isSignedIn}`);
+          const suffix = this._canSignIn() ? 'eligible' : 'ineligible';
```
Comment (vladikoff) [inline app/scripts/views/connect_another_device.js]: I think we try to stay away from ternary in our code base

Comment (shane-tomlinson) [inline app/scripts/views/connect_another_device.js]: > I think we try to stay away from ternary in our code base

I didn't realize that, I personally don't mind non-nested ternary operators. OTOH, nested ones are just hard to read. I'm neither for nor against banning ternary, but we decide it's banned, let's add the [no-ternary eslint rule](http://eslint.org/docs/rules/no-ternary) so we can move on to thinking about more important things.

File: app/scripts/lib/constants.js (hunk 4 above)
Comment (shane-tomlinson) [inline app/scripts/lib/constants.js]: I just noticed a lot of this link is duplicated elsewhere. Perhaps what we should do is extract the the common parts into constants.js and append things like `campaign` that can change depending on context. Sound reasonable?

File: app/scripts/views/complete_sign_up.js [hunk 8]
```diff
@@ -70,89 +66,114 @@ define(function (require, exports, module) {
...
+      return this.invokeBrokerMethod('afterCompleteSignUp', account)
+        .then(() => this._navigateToNextScreen());
+    },
+
+    _navigateToNextScreen () {
+      var account = this.getAccount();
+      var relier = this.relier;
+
+      if (relier.isSync()) {
+        if (this.isInExperimentGroup('connectAnotherDevice', 'treatment')) {
```
Comment (vladikoff) [inline app/scripts/views/complete_sign_up.js]: 2 things we need here, a way to force the treatment / control behaviour.
and a PR in https://github.com/mozilla/fxa-content-experiments/pulls to test this with.

**Comment by vladikoff** [comment]
Dialog text seems not centered:

File: app/scripts/templates/connect_another_device.mustache [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    <div class="graphic graphic-connect-another-device">{{#t}}Success{{/t}}</div>
+
+    {{#canSignIn}}
+      <p>
+        {{#t}}Complete set-up by signing in to this Firefox.{{/t}}
+      </p>
+
+      <form novalidate>
+        <div class="button-row">
+          <a id="signin" class="button" href="{{{ escapedSignInUrl }}}">{{#unsafeTranslate}}Sign in &rarr;{{/unsafeTranslate}}</a>
```
Comment (vladikoff) [inline app/scripts/templates/connect_another_device.mustache]: The metrics are not flushing here (unless I'm missing something).

File: docs/query-params.md [hunk 10]
```diff
@@ -337,6 +337,22 @@ If the "mailcheck" feature should be used.
...
+    encourages users to set up multiple devices.
+* `mailcheck` - provide a tooltip to correct common domain name errors in the
+    email.
+
+### `forceExperimentGroup`
+Force the user into a particular AB test experiment group.
+
+#### Options
+* `control` - default behavior.
+* `treatment` - new behavior.
```
Comment (vladikoff) [inline docs/query-params.md]: we also need to add the new metrics events into docs

File: app/scripts/templates/connect_another_device.mustache [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+          {{#t}}Complete set-up by connecting Firefox for your smartphone.{{/t}}
+        </p>
+      {{/isOther}}
+
+      {{^isFirefoxIos}}
+      <div class="marketing-area"></div>
+      {{/isFirefoxIos}}
+    {{/canSignIn}}
+    <div class="links">
+      <a href="/connect_another_device/why">{{#t}}Why is this required?{{/t}}</a>
```
Comment (vladikoff) [inline app/scripts/templates/connect_another_device.mustache]: need metrics to see if anyone actually clicks on this

File: app/scripts/views/connect_another_device.js [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+    context () {
+      const email = this._getEmail();
+      const signInContext = this._getSignInContext();
+      const escapedSignInUrl = this._getEscapedSignInUrl(signInContext, email);
+      const uap = this._getUap();
+      const isAndroid = uap.isAndroid();
+      const isIos = uap.isIos();
+      const isFirefoxIos = isIos && uap.isFirefox();
+      const isOtherIos = isIos && ! isFirefoxIos;
+      const isOther = ! isAndroid && ! isIos;
```
Comment (vladikoff) [inline app/scripts/views/connect_another_device.js]: might be useful to get metrics of the most popular flow, isAndroid vs others

**Comment by vladikoff** [comment]
Weird issue when i spoof as Nexus 5x clickin on the Google Play link does not work in FF or Chrome: 

UA: `Mozilla/5.0 (Linux; Android 6.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/55.0.2883.95 Mobile Safari/537.36`

File: app/scripts/lib/experiments/connect-another-device.js [hunk 13]
```diff
@@ -0,0 +1,28 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+define(function (require, exports, module) {
```
Comment (vladikoff) [inline app/scripts/lib/experiments/connect-another-device.js]: Nit: add a doc for the future about what this experiment does...

File: app/scripts/views/complete_sign_up.js (hunk 8 above)
Comment (shane-tomlinson) [inline app/scripts/views/complete_sign_up.js]: We have a way to force treatment/control behavior - the tests use it already - https://github.com/mozilla/fxa-content-server/pull/4370/files#diff-4b30e0b3783b9ee05813174ab98c3916R85

experiment PR - https://github.com/mozilla/fxa-content-experiments/pull/49

File: app/scripts/templates/connect_another_device.mustache (hunk 11 above)
Comment (shane-tomlinson) [inline app/scripts/templates/connect_another_device.mustache]: It'll be logged as a screen event.

File: app/scripts/templates/connect_another_device.mustache (hunk 9 above)
Comment (shane-tomlinson) [inline app/scripts/templates/connect_another_device.mustache]: Very strange, it did for me:

> content-server PORT 3030-3 fxa-content-server.server.requests.INFO: POST /metrics 200 27.231
content-server PORT 3030-3 {"time":"2016-12-15T10:00:00.000Z","op":"client.metrics","hostname":"Shanes-Retina-MBP.local","pid":45211,"v":1,"lang":"en","agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.12; rv:53.0) Gecko/20100101 Firefox/53.0","duration":4735,"context":"web","broker":"base","entrypoint":"none","service":"none","migration":"none","nt.navigationStart":0,"nt.unloadEventStart":1808,"nt.unloadEventEnd":1808,"nt.fetchStart":1449,"nt.domainLookupStart":1449,"nt.domainLookupEnd":1449,"nt.connectStart":1449,"nt.connectEnd":1449,"nt.requestStart":1450,"nt.responseStart":1806,"nt.responseEnd":1806,"nt.domLoading":6,"nt.domInteractive":28,"nt.domContentLoadedEventStart":29,"nt.domContentLoadedEventEnd":30,"nt.domComplete":1878,"nt.loadEventStart":1878,"nt.loadEventEnd":1878,"events":["experiment.treatment.mailcheck.enrolled","screen.connect-another-device","loaded"],"event_durations":[2627,2643,2646],"screen.width":1680,"screen.height":1050}

File: app/scripts/templates/connect_another_device.mustache (hunk 11 above)
Comment (shane-tomlinson) [inline app/scripts/templates/connect_another_device.mustache]: screen shot 2016-12-15 at 10 49 05

File: app/scripts/templates/connect_another_device.mustache (hunk 9 above)
Comment (shane-tomlinson) [inline app/scripts/templates/connect_another_device.mustache]: This anchor is considered an external link since it starts with `https`, and [all external links cause a metrics flush](https://github.com/mozilla/fxa-content-server/blob/a54eac025f0d33e399310cf8ba6b88dd11ea0016/app/scripts/views/mixins/external-links-mixin.js#L70). Since the [external links mixin is mixed into the BaseView](https://github.com/mozilla/fxa-content-server/blob/a54eac025f0d33e399310cf8ba6b88dd11ea0016/app/scripts/views/base.js#L1001:L1004), all links that start with `http` automatically have this behavior.

Comment (shane-tomlinson) [inline app/scripts/templates/connect_another_device.mustache]: Maybe we should log a click on these external links so we know where users are exiting. That info doesn't seem to be logged anywhere.

File: app/scripts/lib/constants.js [hunk 14]
```diff
@@ -112,8 +112,13 @@ define(function (require, exports, module) {
     // Login delay for iOS broker
     IOS_V1_LOGIN_MESSAGE_DELAY_MS: 10000,
 
-    UNBLOCK_CODE_LENGTH: 8
+    UNBLOCK_CODE_LENGTH: 8,
+
+    MARKETING_ID_SPRING_2015: 'spring-2015-android-ios-sync',
+    MARKETING_ID_AUTUMN_2016: 'autumn-2016-connect-another-device',
+
+    DOWNLOAD_LINK_ANDROID: 'https://app.adjust.com/2uo1qc?campaign=%(campaign)s&creative=%(creative)s&adgroup=android',
```
Comment (shane-tomlinson) [inline app/scripts/lib/constants.js]: Add a `_TEMPLATE` suffix to the download links so future devs know they need to do some interpolation.

**Comment by vladikoff** [comment]
> Weird issue when i spoof as Nexus 5x clickin on the Google Play link does not work

Ah this is probably not working because Adjust tries to navigate to Native Google play App, so should be fine when on a real device 👍

File: app/scripts/templates/connect_another_device.mustache (hunk 9 above)
Comment (shane-tomlinson) [inline app/scripts/templates/connect_another_device.mustache]: > Maybe we should log a click on these external links so we know where users are exiting. 

[The experiment logs clicks on the signin button](https://github.com/mozilla/fxa-content-server/pull/4370/files#diff-6c80c7633a955dc6fcaf7c93c216faf8R13). Good to go!

File: app/scripts/views/connect_another_device.js (hunk 12 above)
Comment (shane-tomlinson) [inline app/scripts/views/connect_another_device.js]: We kind of sort of get this information via the "marketing impression" counts, but I think you are right, we should make this explicit.

File: app/scripts/templates/connect_another_device.mustache (hunk 11 above)
Comment (vladikoff) [inline app/scripts/templates/connect_another_device.mustache]: awesome, that should be good enough 👍

**Comment by vladikoff** [comment]
@shane-tomlinson on a 'tablet' resolution I noticed that the modal gets the 100% width:

File: app/scripts/templates/marketing_snippet.mustache [hunk 15]
```diff
@@ -28,13 +29,30 @@
...
     </aside>
-  {{/showSignUpMarketing}}
+  {{/isSpring2015}}
+  {{#isAutumn2016}}
+    {{#isIos}}
+      <a class="marketing-link marketing-link-ios" href="{{ downloadLinkIos }}" data-marketing-id="{{ marketingId }}" data-marketing-type="ios"><img width="152" height="53" src="{{ appStoreImage }}" alt="{{#t}}Download Firefox for iOS!{{/t}}" /></a>
+    {{/isIos}}
+
+    {{#isAndroid}}
+      <a class="marketing-link marketing-link-android" href="{{ downloadLinkAndroid }}" data-marketing-id="{{ marketingId }}" data-marketing-type="android"><img width="152" height="45" src="{{ playStoreImage }}" alt="{{#t}}Download Firefox for Android.{{/t}}"></a>
```
Comment (vladikoff) [inline app/scripts/templates/marketing_snippet.mustache]: the android `img` tag is not `/>` like the iOS one

File: app/scripts/templates/marketing_snippet.mustache [hunk 16]
```diff
@@ -28,13 +29,30 @@
...
+    {{/isAndroid}}
+
+    {{#isOther}}
+      <div class="links cf">
+        <a class="marketing-link marketing-link-ios left" href="{{ downloadLinkIos }}" data-marketing-id="{{ marketingId }}" data-marketing-type="ios"><img width="152" height="45" src="{{ appStoreImage }}" alt="{{#t}}Download Firefox for iOS!{{/t}}" /></a>
+        <a class="marketing-link marketing-link-android right" href="{{ downloadLinkAndroid }}" data-marketing-id="{{ marketingId }}" data-marketing-type="android"><img width="152" height="45"  src="{{ playStoreImage }}" alt="{{#t}}Download Firefox for Android.{{/t}}"></a>
```
Comment (vladikoff) [inline app/scripts/templates/marketing_snippet.mustache]: same as https://github.com/mozilla/fxa-content-server/pull/4370/files#r92641114

File: app/scripts/templates/why_connect_another_device.mustache [hunk 17]
```diff
@@ -0,0 +1,13 @@
+
```
Comment (vladikoff) [inline app/scripts/templates/why_connect_another_device.mustache]: nit: extra line break?

File: app/scripts/templates/why_connect_another_device.mustache [hunk 18]
```diff
@@ -0,0 +1,13 @@
...
+<aside class="modal-panel connect-another-device">
+  <header>
+    <h1 id="fxa-why-connect-another-device-header">Why Sync two (or more!) devices?</h1>
+  </header>
+
+  <p>
+    {{#t}}Sync lets you share the bookmarks, history and passwords you save to Firefox on one device with Firefox on other devices. Sign in to Firefox on your phone or tablet and everything you've saved to Firefox for desktop will appear.{{/t}}
+  </p>
+  <p>
+    {{#t}}Sync isn't designed as a permanent backup – it's just sharing what is saved locally. If you reset your password and your data isn't stored on at least one other device, you could lose all the information associated with the old password. That's why you'll want to install Firefox and use Sync on more than one device.{{/t}}
```
Comment (vladikoff) [inline app/scripts/templates/why_connect_another_device.mustache]: nit: should be line break so it is easier to read the source?

File: app/scripts/views/marketing_snippet.js [hunk 19]
```diff
@@ -130,6 +143,20 @@ define(function (require, exports, module) {
...
+     * Add the appropriate metrics related query params to
+     * the Firefox app's app store link.
+     *
+     * @param {String} base
+     * @returns {String}
+     */
+    _storeLink (base) {
+      return Strings.interpolate(base, {
+        campaign: 'fxa-conf-page',
+        creative: View.BUTTON_IDS[ this._marketingId ],
```
Comment (vladikoff) [inline app/scripts/views/marketing_snippet.js]: nit: the spaces here and below are a bit strange ( `[ this._marketingId ]`) . A bit of inconsistency with the rest of the code base. example, we have `var command = this.commands[commandName];` in fx-sync.

File: app/scripts/templates/marketing_snippet.mustache [hunk 20]
```diff
@@ -28,13 +29,30 @@
...
+    {{#isIos}}
+      <a class="marketing-link marketing-link-ios" href="{{ downloadLinkIos }}" data-marketing-id="{{ marketingId }}" data-marketing-type="ios"><img width="152" height="53" src="{{ appStoreImage }}" alt="{{#t}}Download Firefox for iOS!{{/t}}" /></a>
+    {{/isIos}}
+
+    {{#isAndroid}}
+      <a class="marketing-link marketing-link-android" href="{{ downloadLinkAndroid }}" data-marketing-id="{{ marketingId }}" data-marketing-type="android"><img width="152" height="45" src="{{ playStoreImage }}" alt="{{#t}}Download Firefox for Android.{{/t}}"></a>
+    {{/isAndroid}}
+
+    {{#isOther}}
+      <div class="links cf">
```
Comment (shane-tomlinson) [inline app/scripts/templates/marketing_snippet.mustache]: clearfix. Needed to prevent the "Why is this required" text from showing up smushed between the links to the stores since they both use float.

File: app/scripts/templates/why_connect_another_device.mustache (hunk 17 above)
Comment (shane-tomlinson) [inline app/scripts/templates/why_connect_another_device.mustache]: Yeah, that's how it is in almost all the other templates after the `header` too.

File: app/scripts/views/base.js [hunk 21]
```diff
@@ -931,12 +931,24 @@ define(function (require, exports, module) {
...
+     * @returns {String}
+     */
+    getSearchParam (paramName) {
+      return Url.searchParam(paramName, this.window.location.search);
+    },
+
+    /**
+     * Get values from the URL search parameters.
+     *
+     * @param {String[]} paramNames - name of the search parameters
```
Comment (shane-tomlinson) [inline app/scripts/views/base.js]: `paramNames` is optional, mark it as such.

File: app/scripts/views/complete_sign_up.js [hunk 22]
```diff
@@ -70,89 +66,114 @@ define(function (require, exports, module) {
     },
 
     beforeRender () {
-      var verificationInfo = this._verificationInfo;
+      const verificationInfo = this._verificationInfo;
```
Comment (shane-tomlinson) [inline app/scripts/views/complete_sign_up.js]: extract the majority of this cleanup into its own PR to make each PR simpler to review.

File: app/scripts/views/connect_another_device.js [hunk 23]
```diff
@@ -0,0 +1,269 @@
...
+    }
+  }, {
+    ENTRYPOINT: 'connect_another_device'
+  });
+
+  Cocktail.mixin(
+    View,
+    ExperimentMixin,
+    MarketingMixin({
+      autocreate: false
```
Comment (shane-tomlinson) [inline app/scripts/views/connect_another_device.js]: why no autocreate?

File: app/scripts/views/connect_another_device.js [hunk 24]
```diff
@@ -0,0 +1,269 @@
...
+    /**
+     * Return the sign in context that can be used to sign in
+     * to Sync. Assumes the context is only used if the user
+     * can actually sign in to Sync.
+     *
+     * @returns {String}
+     * @private
+     */
+    _getSignInContext() {
+      if (this._getUap().isAndroid()) {
```
Comment (shane-tomlinson) [inline app/scripts/views/connect_another_device.js]: Change `isAndroid` to `isFirefoxAndroid`

File: app/scripts/views/connect_another_device.js [hunk 25]
```diff
@@ -0,0 +1,269 @@
...
+        return Constants.FX_FENNEC_V1_CONTEXT;
+      } else {
```
Comment (shane-tomlinson) [inline app/scripts/views/connect_another_device.js]: Add a `isFirefox() && ! isFirefoxIos()` check here to avoid making assumptions about where the output of this function will be used.

File: app/scripts/views/connect_another_device.js [hunk 26]
```diff
@@ -0,0 +1,269 @@
...
+      return `${origin}/signin${escapedSearchString}`;
+    },
+
+    /**
+     * Get the current user's email address
+     *
+     * @returns {String}
+     * @private
+     */
+    _getEmail () {
```
Comment (shane-tomlinson) [inline app/scripts/views/connect_another_device.js]: `_getEmail` is only used in one place, is it needed?

File: app/scripts/views/marketing_snippet.js [hunk 27]
```diff
@@ -60,8 +64,20 @@ define(function (require, exports, module) {
   const View = BaseView.extend({
     template: Template,
 
+    /**
+     * Initialize.
+     *
+     * @param {Object} [options=}|]
```
Comment (shane-tomlinson) [inline app/scripts/views/marketing_snippet.js]: `options=}|`=>`options={}`

File: app/scripts/views/marketing_snippet.js [hunk 28]
```diff
@@ -60,8 +64,20 @@ define(function (require, exports, module) {
...
+     *  @param {String} [options.marketingId] - marketing ID to use for logging.
```
Comment (shane-tomlinson) [inline app/scripts/views/marketing_snippet.js]: Many of the changes to marketing_snippet.js could be extracted into its own PR to simplify review. For example, the marketingId and which options could be extracted, but unused, against master.

File: app/scripts/views/complete_sign_up.js [hunk 29]
```diff
@@ -70,89 +66,114 @@ define(function (require, exports, module) {
...
-            return account.isSignedIn()
-              .then((isSignedIn) => {
-                if (isSignedIn) {
-                  this.navigate('settings', {
-                    success: t('Account verified successfully')
-                  });
-                } else {
-                  this._navigateToVerifiedScreen();
-                }
+      return this.user.completeAccountSignUp(account, code, options)
```
Comment (shane-tomlinson) [inline app/scripts/views/complete_sign_up.js]: Done

**Comment by shane-tomlinson** [comment]
> @shane-tomlinson on a 'tablet' resolution I noticed that the modal gets the 100% width:
> not sure if we need to change any of that... 

What is the resolution? It looks like it's hitting the "@media (max-width: 520px), (orientation: landscape) and (min-width: 481px) and (max-height: 480px) {" media query which is meant for mobile phones in landscape mode, that's how the `@include repond-to('small')` mixin works.

I'm not sure what the best approach here is. I'm tempted to say "that's fine for mobile." Throwing this PR up on stomlinson.dev.lcip.org to try on my iPad 2 to see how it responds on a real tablet in landscape mode.

**Comment by vladikoff** [comment]
Nice! 👍
